include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,169 @@
+#ifndef KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+#include "kokkos-utils/timer/Timer.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+template <typename...>
+class TimerListener;
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The start event is selected by requiring it to be of type @p BeginEventType and
+ * to be a match for the passed matcher. The stop event must be of type @p EndEventType
+ * and have the same ID as the begin event.
+ *
+ * The underlying timer @ref timer is a @ref Kokkos::utils::timer::Timer, instantiated
+ * for the execution space @p Exec. This class stores the execution space
+ * instance passed to the constructor, and uses it as the one passed to @ref timer.
+ *
+ * This class is well-suited for timing the execution of a parallel workload, such as a
+ * @c Kokkos parallel for, parallel reduce, parallel scan, or deep copy operation, to which
+ * an execution space instance is passed. In such a situation, the execution space instance
+ * passed to this class's constructor should be the same as the one passed to the workload.
+ */
+template <
+    Kokkos::utils::callbacks::BeginEvent BeginEventType, Kokkos::utils::callbacks::EndEvent EndEventType,
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<BeginEventType>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<BeginEventType, EndEventType, MatcherType, Exec>
+{
+public:
+    static constexpr uint64_t invalid_event_id = Kokkos::Experimental::finite_max_v<uint64_t>;
+
+public:
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
```

Comment:
I guess it is not needed and in fact will constrain the constructor too much ? What if my matcher can be constructed from *e.g.* a `std::string` ?
```suggestion
    template <typename T>
```

Reply:
This is a good suggestion indeed, because there's quite some duplication and long lines of code now. We haven't done it for other similar cases. Probably best to do it only for the cases in this PR.

Reply:
OK. Done indeed as said in an earlier thread. Done everywhere.